File: js/plugins/firebase/tests/context_test.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, expect, it } from '@jest/globals';
import type { RequestData } from 'genkit/context';
import {
  fakeToken,
  firebaseContext,
  setDebugSkipTokenVerification,
  type FirebaseContext,
} from '../src/context';

function request(headers: Record<string, string> = {}): RequestData {
  return {
    method: 'POST',
    headers,
    input: undefined,
  };
}
describe('firebaseAuth', () => {
  setDebugSkipTokenVerification(true);

  describe('no policy', () => {
    it('handles noop', async () => {
      const context = await firebaseContext()(request());
      expect(context).toEqual({});
    });

    it('handles all headers', async () => {
      const context = await firebaseContext()(
        request({
          authorization: `bearer ${fakeToken({ sub: 'user' })}`,
          'x-firebase-appcheck': fakeToken({ sub: 'appId' }),
          'firebase-instance-id-token': 'token',
        })
      );
      expect(context).toEqual({
        auth: {
          uid: 'user',
          token: { sub: 'user' },
          rawToken: fakeToken({ sub: 'user' }),
        },
        app: {
          appId: 'appId',
          token: { sub: 'appId' },
          alreadyConsumed: false,
          rawToken: fakeToken({ sub: 'appId' }),
        },
        instanceIdToken: 'token',
      });
    });
  });

  describe('declaritive policies', () => {
    it('handles signedIn', async () => {
      expect(() =>
        firebaseContext({ signedIn: true })(request())
      ).rejects.toThrow();
      expect(await firebaseContext({ signedIn: false })(request())).toEqual({});
      expect(
        await firebaseContext({ signedIn: false })(
          request({ authorization: `bearer ${fakeToken({ sub: 'user' })}` })
        )
      ).toEqual({
        auth: {
          uid: 'user',
          token: {
            sub: 'user',
          },
          rawToken: fakeToken({ sub: 'user' }),
        },
      });
    });

    it('handles emailVerified', async () => {
      expect(() =>
        firebaseContext({ emailVerified: true })(request())
      ).rejects.toThrow();
      expect(
        await firebaseContext({ emailVerified: false })(request())
      ).toEqual({});
      expect(() =>
        firebaseContext({ emailVerified: true })(
          request({ authorization: `bearer ${fakeToken({ sub: 'user' })}` })
        )
      ).rejects.toThrow();
      expect(() =>
        firebaseContext({ emailVerified: true })(
          request({
            authorization: `bearer ${fakeToken({ sub: 'user', email: 'user@google.com', email_verified: 'false' })}`,
          })
        )
      ).rejects.toThrow();
      expect(
        await firebaseContext({ emailVerified: true })(
          request({
            authorization: `bearer ${fakeToken({ sub: 'user', email: 'user@google.com', email_verified: 'true' })}`,
          })
        )
      ).toEqual({
        auth: {
          uid: 'user',
          token: {
            sub: 'user',
            email: 'user@google.com',
            email_verified: 'true',
          },
          rawToken: fakeToken({
            sub: 'user',
            email: 'user@google.com',
            email_verified: 'true',
          }),
        },
      });
    });

    it('enforces hasClaim (string)', async () => {
      expect(
        await firebaseContext({ hasClaim: 'email' })(
          request({
            authorization: `bearer ${fakeToken({ sub: 'user', email: 'user@google.com' })}`,
          })
        )
      ).toEqual({
        auth: {
          uid: 'user',
          token: {
            sub: 'user',
            email: 'user@google.com',
          },
          rawToken: fakeToken({ sub: 'user', email: 'user@google.com' }),
        },
      });
      expect(() =>
        firebaseContext({ hasClaim: 'admin' })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
            })}`,
          })
        )
      ).rejects.toThrow();
      expect(() =>
        firebaseContext({ hasClaim: 'admin' })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
              admin: 'false',
            })}`,
          })
        )
      ).rejects.toThrow();
    });

    it('handles hasClaim(string[])', async () => {
      expect(() =>
        firebaseContext({ hasClaim: ['email', 'admin'] })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
              admin: 'true',
            })}`,
          })
        )
      ).rejects.toThrow();
      expect(() =>
        firebaseContext({ hasClaim: ['email', 'admin'] })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
              admin: 'false',
              email: 'user@google.com',
            })}`,
          })
        )
      ).rejects.toThrow();
      expect(
        await firebaseContext({ hasClaim: ['email', 'admin'] })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
              admin: 'true',
              email: 'user@google.com',
            })}`,
          })
        )
      ).toEqual({
        auth: {
          uid: 'user',
          token: {
            sub: 'user',
            admin: 'true',
            email: 'user@google.com',
          },
          rawToken: fakeToken({
            sub: 'user',
            admin: 'true',
            email: 'user@google.com',
          }),
        },
      });
    });

    it('handles hasClaim(Record<string, boolean>)', async () => {
      expect(() =>
        firebaseContext({
          hasClaim: {
            admin: 'true',
            humor: 'dad',
          },
        })(
          request({
            authorization: `bearer ${fakeToken({
              uid: 'user',
              admin: 'true',
            })}`,
          })
        )
      ).rejects.toThrow();
      expect(() =>
        firebaseContext({
          hasClaim: {
            admin: 'true',
            humor: 'dad',
          },
        })(
          request({
            authorization: `bearer ${fakeToken({
              uid: 'user',
              admin: 'true',
              humor: 'programming',
            })}`,
          })
        )
      ).rejects.toThrow();
      expect(
        await firebaseContext({
          hasClaim: {
            admin: 'true',
            humor: 'dad',
          },
        })(
          request({
            authorization: `bearer ${fakeToken({
              sub: 'user',
              admin: 'true',
              humor: 'dad',
            })}`,
          })
        )
      ).toEqual({
        auth: {
          uid: 'user',
          token: {
            sub: 'user',
            admin: 'true',
            humor: 'dad',
          },
          rawToken: fakeToken({
            sub: 'user',
            admin: 'true',
            humor: 'dad',
          }),
        },
      });
    });

    it('handles enforceAppCheck', async () => {
      expect(
        await firebaseContext({ enforceAppCheck: true })(
          request({
            'x-firebase-appcheck': fakeToken({ sub: 'appId' }),
          })
        )
      ).toEqual({
        app: {
          appId: 'appId',
          token: {
            sub: 'appId',
          },
          alreadyConsumed: false,
          rawToken: fakeToken({ sub: 'appId' }),
        },
      });
      expect(() =>
        firebaseContext({ enforceAppCheck: true })(request())
      ).rejects.toThrow();
    });
  });

  describe('policy functions', () => {
    it('passes context', () => {
      firebaseContext((context: FirebaseContext) => {
        expect(context).toEqual({
          auth: {
            uid: 'user',
            token: {
              sub: 'user',
            },
            rawToken: fakeToken({ sub: 'user' }),
          },
          app: {
            appId: 'app',
            token: {
              sub: 'app',
            },
            alreadyConsumed: false,
            rawToken: fakeToken({ sub: 'app' }),
          },
          instanceIdToken: 'iid',
        });
      })(
        request({
          authorization: `bearer ${fakeToken({ sub: 'user' })}`,
          'x-firebase-appcheck': fakeToken({ sub: 'app' }),
          'firebase-instance-id-token': 'iid',
        })
      );
      firebaseContext((context: FirebaseContext) => {
        expect(context).toEqual({});
      })(request());
    });
  });
});

File: py/packages/genkit/tests/core/context_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tests for genkit.core.context — api_key() context provider.

These tests mirror the JS SDK's core/tests/context_test.ts to ensure
behavioral parity across SDKs.
"""

from typing import Any

import pytest

from genkit.core.context import ApiKeyContext, api_key
from genkit.core.error import UserFacingError


def _request(key: str | None = None) -> dict[str, Any]:
    """Build a minimal request_data dict matching the flows server format.

    Args:
        key: Optional Authorization header value.

    Returns:
        Dict with method, headers, and input fields.
    """
    headers: dict[str, str] = {}
    if key is not None:
        headers['authorization'] = key
    return {
        'method': 'POST',
        'headers': headers,
        'input': {},
    }


# The flows server passes (context, request_data) to providers.
_EMPTY_CTX: dict[str, Any] = {}


class TestApiKeyPassthrough:
    """Tests for api_key() in pass-through mode (no argument)."""

    @pytest.mark.asyncio
    async def test_no_key_returns_none(self) -> None:
        """Pass-through with no Authorization header returns api_key=None."""
        provider = api_key()
        result = await provider(_EMPTY_CTX, _request())
        assert result == {'auth': {'api_key': None}}

    @pytest.mark.asyncio
    async def test_with_key_returns_key(self) -> None:
        """Pass-through with Authorization header returns the key."""
        provider = api_key()
        result = await provider(_EMPTY_CTX, _request('my-key'))
        assert result == {'auth': {'api_key': 'my-key'}}


class TestApiKeyExactMatch:
    """Tests for api_key('expected') in exact-match mode."""

    @pytest.mark.asyncio
    async def test_correct_key_succeeds(self) -> None:
        """Exact match with correct key returns the key."""
        provider = api_key('secret')
        result = await provider(_EMPTY_CTX, _request('secret'))
        assert result == {'auth': {'api_key': 'secret'}}

    @pytest.mark.asyncio
    async def test_wrong_key_raises_permission_denied(self) -> None:
        """Exact match with wrong key raises PERMISSION_DENIED."""
        provider = api_key('secret')
        with pytest.raises(UserFacingError, match='Permission Denied') as exc_info:
            await provider(_EMPTY_CTX, _request('wrong-key'))
        assert exc_info.value.status == 'PERMISSION_DENIED'

    @pytest.mark.asyncio
    async def test_missing_key_raises_unauthenticated(self) -> None:
        """Exact match with no key raises UNAUTHENTICATED."""
        provider = api_key('secret')
        with pytest.raises(UserFacingError, match='Unauthenticated') as exc_info:
            await provider(_EMPTY_CTX, _request())
        assert exc_info.value.status == 'UNAUTHENTICATED'

    @pytest.mark.asyncio
    async def test_empty_key_raises_unauthenticated(self) -> None:
        """Exact match with empty string key raises UNAUTHENTICATED."""
        provider = api_key('secret')
        with pytest.raises(UserFacingError, match='Unauthenticated') as exc_info:
            await provider(_EMPTY_CTX, _request(''))
        assert exc_info.value.status == 'UNAUTHENTICATED'


class TestApiKeyCustomPolicy:
    """Tests for api_key(policy_fn) in custom policy mode."""

    @pytest.mark.asyncio
    async def test_sync_policy_receives_context(self) -> None:
        """Sync policy function receives the correct ApiKeyContext."""
        captured: list[ApiKeyContext] = []

        def capture_policy(ctx: ApiKeyContext) -> None:
            captured.append(ctx)

        provider = api_key(capture_policy)
        result = await provider(_EMPTY_CTX, _request('test-key'))

        assert len(captured) == 1
        assert captured[0].auth == {'api_key': 'test-key'}
        assert result == {'auth': {'api_key': 'test-key'}}

    @pytest.mark.asyncio
    async def test_sync_policy_with_no_key(self) -> None:
        """Sync policy receives api_key=None when no header present."""
        captured: list[ApiKeyContext] = []

        def capture_policy(ctx: ApiKeyContext) -> None:
            captured.append(ctx)

        provider = api_key(capture_policy)
        await provider(_EMPTY_CTX, _request())

        assert len(captured) == 1
        assert captured[0].auth == {'api_key': None}

    @pytest.mark.asyncio
    async def test_sync_policy_can_raise(self) -> None:
        """Sync policy can reject requests by raising."""

        def deny_all(ctx: ApiKeyContext) -> None:
            raise UserFacingError('PERMISSION_DENIED', 'Denied')

        provider = api_key(deny_all)
        with pytest.raises(UserFacingError, match='Denied'):
            await provider(_EMPTY_CTX, _request('any-key'))

    @pytest.mark.asyncio
    async def test_async_policy_receives_context(self) -> None:
        """Async policy function receives the correct ApiKeyContext."""
        captured: list[ApiKeyContext] = []

        async def capture_policy(ctx: ApiKeyContext) -> None:
            captured.append(ctx)

        provider = api_key(capture_policy)
        result = await provider(_EMPTY_CTX, _request('async-key'))

        assert len(captured) == 1
        assert captured[0].auth == {'api_key': 'async-key'}
        assert result == {'auth': {'api_key': 'async-key'}}

    @pytest.mark.asyncio
    async def test_async_policy_can_raise(self) -> None:
        """Async policy can reject requests by raising."""

        async def deny_all(ctx: ApiKeyContext) -> None:
            raise UserFacingError('UNAUTHENTICATED', 'Must authenticate')

        provider = api_key(deny_all)
        with pytest.raises(UserFacingError, match='Must authenticate'):
            await provider(_EMPTY_CTX, _request('any-key'))


class TestApiKeyContext:
    """Tests for the ApiKeyContext dataclass."""

    def test_equality(self) -> None:
        """Two ApiKeyContext with the same key are equal."""
        ctx1 = ApiKeyContext('key')
        ctx2 = ApiKeyContext('key')
        assert ctx1 == ctx2

    def test_inequality(self) -> None:
        """Two ApiKeyContext with different keys are not equal."""
        ctx1 = ApiKeyContext('key1')
        ctx2 = ApiKeyContext('key2')
        assert ctx1 != ctx2

    def test_none_key(self) -> None:
        """ApiKeyContext with None key is valid."""
        ctx = ApiKeyContext(None)
        assert ctx.auth == {'api_key': None}
        assert ctx.to_dict() == {'auth': {'api_key': None}}

    def test_repr(self) -> None:
        """ApiKeyContext repr includes the auth dict."""
        ctx = ApiKeyContext('my-key')
        assert 'my-key' in repr(ctx)
        assert 'ApiKeyContext' in repr(ctx)

    def test_to_dict(self) -> None:
        """to_dict returns a plain dict suitable for context merging."""
        ctx = ApiKeyContext('k')
        assert ctx.to_dict() == {'auth': {'api_key': 'k'}}

    def test_inequality_with_other_type(self) -> None:
        """ApiKeyContext is not equal to a non-ApiKeyContext object."""
        ctx = ApiKeyContext('key')
        assert ctx != 'not-a-context'
